extensions/kubernetes-client/deployment-internal/src/main/java/io/quarkus/kubernetes/client/deployment/DevServicesKubernetesProcessor.java

```diff
@@ -0,0 +1,415 @@
+            this.shared = config.shared;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(devServicesEnabled, flavor, apiVersion, overrideKubeconfig, shared, serviceName);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj)
+                return true;
+            if (!(obj instanceof KubernetesDevServiceCfg))
+                return false;
+            KubernetesDevServiceCfg other = (KubernetesDevServiceCfg) obj;
+            return devServicesEnabled == other.devServicesEnabled && flavor == other.flavor
+                    && Objects.equals(apiVersion, other.apiVersion) && overrideKubeconfig == other.overrideKubeconfig
+                    && shared == other.shared && Objects.equals(serviceName, other.serviceName);
+        }
+    }
+
+    private class RunningContainer implements ContainerState {
+        private static final String KIND_KUBECONFIG = "/etc/kubernetes/admin.conf";
+        private static final String K3S_KUBECONFIG = "/etc/rancher/k3s/k3s.yaml";
+        private static final String APISERVER = "apiserver";
+        private static final String PKI_BASEDIR = "/etc/kubernetes/pki";
+        private static final String API_SERVER_CA = PKI_BASEDIR + "/ca.crt";
+        private static final String API_SERVER_CERT = PKI_BASEDIR + "/apiserver.crt";
+        private static final String API_SERVER_KEY = PKI_BASEDIR + "/apiserver.key";
+
+        private final DockerClient dockerClient;
+
+        private final InspectContainerResponse containerInfo;
+
+        private final ContainerAddress containerAddress;
+
+        public RunningContainer(DockerClient dockerClient, ContainerAddress containerAddress) {
+            this.dockerClient = dockerClient;
+            this.containerAddress = containerAddress;
+            this.containerInfo = dockerClient.inspectContainerCmd(getContainerId()).exec();
+        }
+
+        public Map<String, String> getKubeconfig() {
+            var image = getContainerInfo().getConfig().getImage();
+            if (image.contains("rancher/k3s")) {
+                return getKubernetesClientConfigFromKubeConfig(
+                        KubeConfigUtils.parseKubeConfig(KubeConfigUtils.replaceServerInKubeconfig(containerAddress.getUrl(),
+                                getFileContentFromContainer(K3S_KUBECONFIG))));
+            } else if (image.contains("kindest/node")) {
+                return getKubernetesClientConfigFromKubeConfig(
+                        KubeConfigUtils.parseKubeConfig(KubeConfigUtils.replaceServerInKubeconfig(containerAddress.getUrl(),
+                                getFileContentFromContainer(KIND_KUBECONFIG))));
+            } else if (image.contains("k8s.gcr.io/kube-apiserver")) {
+                return getKubernetesClientConfigFromKubeConfig(getKubeconfigFromApiContainer(containerAddress.getUrl()));
+            }
+
+            throw new RuntimeException();
```

Comment:
I think this logic should be moved into here: https://github.com/quarkusio/quarkus/pull/28305/files#diff-1f1a987103950b5c6975b74afefa56dea0d2f9a5e592db110493311faf2309abR195

Reply:
@Sgitario I don't see where the link point at.